### PR TITLE
bpo-34270: added the ability to name asyncio tasks

### PR DESCRIPTION
**Pull request by agronholm**
The ability to name individual tasks helps developers debug complex
asyncio applications.

Co-authored-by: Antti Haapala <antti.haapala@anttipatterns.com>

https://bugs.python.org/issue34270

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](https://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


When your account is ready, please add a comment in this pull request
and a Python core developer will remove the `CLA not signed` label
to make the bot check again.

You can [check yourself](https://devguide.python.org/committing.html#checking-if-the-cla-has-been-received)
to see if the CLA has been received.

Thanks again for your contribution, we look forward to reviewing it!

**Comment by agronholm** [comment]
I've updated my account to include my github name, so the CLA label should be cleared.

**Comment by 1st1** [comment]
I'll review this tomorrow

**Comment by ztane** [comment]
Co-author here (CLA signed)

File: Lib/asyncio/tasks.py
Comment (zooba) [inline Lib/asyncio/tasks.py]: Is forcing this to a `str` really important enough to need a property? The name is basically only ever used for printing anyway, which will do the conversion when necessary.

**Review by zooba — APPROVED** [review]
Wait for Yury, but it looks good to me. One question about the need to convert names to str on assignment, but not a big deal.

File: Lib/asyncio/tasks.py
Comment (agronholm) [reply Lib/asyncio/tasks.py]: This is consistent with the behavior of `threading.Thread`. Some code might also access the `name` property directly, expecting a `str`.

File: Lib/asyncio/tasks.py
Comment (elprans) [inline Lib/asyncio/tasks.py]: Using `itertools.count()` seems like a bit of an overkill to me.   I'd simply do an explicit increment.  Also, `_task_name_counter` would be a clearer name.

File: Doc/library/asyncio-eventloop.rst
Comment (1st1) [inline Doc/library/asyncio-eventloop.rst:253]: We also need to update the documentation of `set_task_factory`

File: Doc/library/asyncio-task.rst
Comment (1st1) [inline Doc/library/asyncio-task.rst]: I care way more about consistency of asyncio APIs than about maintaining any kind of artificial similarity between Task and Thread.  And in asyncio we don't use properties that much. Please replace the `name` property with a `Task.get_name()` method.  There's no use case for adding `set_name()` though, so let's not add it at least in 3.8.

File: Lib/asyncio/base_tasks.py
Comment (1st1) [inline Lib/asyncio/base_tasks.py:21]: BTW, feel free to make another pull request to refactor this code. I don't like us modifying some list object with `insert()` calls.

File: Lib/asyncio/tasks.py
Comment (1st1) [reply Lib/asyncio/tasks.py]: Indeed this seems like an overkill. There's no reason to make Task creation any slower than it has to be; keep in mind that PyPy uses the pure Python version. Just replace this with a simple int increment.

File: Lib/asyncio/tasks.py
Comment (1st1) [inline Lib/asyncio/tasks.py]: This is a bit too hard to read. I'd replace it with an `if` statement.

File: Lib/asyncio/tasks.py
Comment (1st1) [reply Lib/asyncio/tasks.py]: As I mentioned above I don't think we need a setter. A `Task.get_name()` function is enough for now.

File: Modules/_asynciomodule.c
Comment (1st1) [inline Modules/_asynciomodule.c]: Please use `uint64_t`.

File: Modules/_asynciomodule.c
Comment (1st1) [inline Modules/_asynciomodule.c]: Please remove this new line to make it easier for the reader to see that we handle the case of `name==NULL`

**Review by 1st1 — CHANGES_REQUESTED** [review]

File: Lib/asyncio/tasks.py
Comment (agronholm) [reply Lib/asyncio/tasks.py]: This is what the threading module does, and it uses `itertools.count()` because `_counter += 1` is not thread safe. The same applies to asyncio if there are more than one thread running an event loop (since the counter is global).

File: Doc/library/asyncio-task.rst
Comment (agronholm) [reply Doc/library/asyncio-task.rst]: Got it.

File: Lib/asyncio/tasks.py
Comment (1st1) [reply Lib/asyncio/tasks.py]: > _counter += 1 is not thread safe.

It is thread safe in Python and will likely always be with or without the GIL.

File: Lib/asyncio/tasks.py
Comment (agronholm) [reply Lib/asyncio/tasks.py]: You can set names of threads and processes too, why not tasks as well? Where's the harm in that? I can think of a few use cases for this, like displaying the status of a task as it progresses.
I could convert this to java-style setters and getters if you want, despite how much I despise them (not to mention the fact that `threading` also switched away from this behavior). But your call.

File: Doc/library/asyncio-eventloop.rst
Comment (agronholm) [reply Doc/library/asyncio-eventloop.rst:253]: Good point.

File: Lib/asyncio/base_tasks.py
Comment (agronholm) [reply Lib/asyncio/base_tasks.py:21]: How would you prefer that we display the name of the task in `repr()` then?

File: Modules/_asynciomodule.c
Comment (agronholm) [reply Modules/_asynciomodule.c]: I had that at first but @ztane pointed out that the format string explicitly requires an `unsigned long long` and `uint64_t` may or may not be the same size. Are you sure this is ok?

File: Lib/asyncio/tasks.py
Comment (1st1) [reply Lib/asyncio/tasks.py]: > I can think of a few use cases for this, like displaying the status of a task as it progresses.

Hm, interesting. OK, that's at least one valid use-case :)

> I could convert this to java-style setters and getters if you want, despite how much I despise them (not to mention the fact that threading also switched away from this behavior). But your call.

It's not about liking Java-style getters or not, it's about consistency.  If asyncio's APIs were using `@property` it'd be equally happy about that, but now we better follow the overall style and use `get_name()` and `set_name()`.

File: Lib/asyncio/base_tasks.py
Comment (1st1) [reply Lib/asyncio/base_tasks.py:21]: It would be great to refactor the code that formats things (tasks, callbacks, reprs) etc somehow. Right now it feels hacky and it's one of the areas of asyncio code that I'd like to be rewritten to be clearer.  Discussing that is slightly out of the scope of this PR though, but if you have any ideas how to avoid mutating lists and having a non-linear formatting logic I'd be happy to accept a new PR.  For this PR, your current code is totally fine!

File: Modules/_asynciomodule.c
Comment (1st1) [reply Modules/_asynciomodule.c]: Got it. AFAIK `unsigned long long` is guaranteed to be large enough to store `uint64_t`. Using the latter makes the code more consistent, so I'd go for it anyways.

You can use this format specifier to format `uint64_t` specifically:

```C
printf("... %" PRIu64 "...", (uint64_t)1);
```

**Comment by 1st1** [comment]
BTW, please also add a What's New entry for 3.8.

**Comment by asvetlov** [comment]
I like the idea!

File: Lib/asyncio/tasks.py
Comment (agronholm) [reply Lib/asyncio/tasks.py]: > It is thread safe in Python and will likely always be with or without the GIL.

What output does [this script](https://gist.github.com/agronholm/c06c2dec5eb77c98ab264151fff6178c) give you then?

Comment (1st1) [reply Lib/asyncio/tasks.py]: Hmmm... thank you for pointing this out! ;)

To my point: `+=` is actually atomic, but storing the updated number in globals is another operation and we have a race between them. This actually deserves a comment in the code explaining why we use `counter()` and not `+=`.

**Comment by agronholm** [comment]
I have made the requested changes; please review again.

**Review by 1st1 — APPROVED** [review]

**Comment by 1st1** [comment]
I was about to hit the "merge" button but decided to quickly take another glance over this PR. And then I realized that this change, strictly speaking, is backwards incompatible.

For example, if there's an asyncio program that supports Python 3.6+ and uses custom task factories, the code will break when asyncio 3.8 will start passing the `name` argument to custom task factories.  Or if someone uses a third-party event loop which isn't updated to have this new feature and they call `asyncio.create_task(coro, name='foo')`.

Therefore I think we need to be extra careful with how exactly we implement this new API.  To mitigate the risks, I propose to do the following:

1. We reinstate `Task.set_name()` API.  You had some compelling arguments for it, but the key argument is that it will allow us to gradually enable this API.

2. `loop.create_task()` should be modified in such a way that it supports existing task factories, which means that it shouldn't pass the `name` argument to them.  So we want to update `loop.create_task` to match the following code:

```python
def Loop.create_task(self, coro, *, name=None):
    """Schedule a coroutine object.

    Return a task object.
    """
    self._check_closed()
    if self._task_factory is None:
        task = tasks.Task(coro, loop=self)
        if task._source_traceback:
            del task._source_traceback[-1]
    else:
        task = self._task_factory(self, coro)

    # --- new code
    if name is not None:
        try:
            set_name = task.set_name
        except AttributeError:
            pass
        else:
            set_name(name)
    # --- end of new code

    return task
```

3. `asyncio.create_task()` should be updated to support third-party event loops that don't have `name` argument in their `loop.create_task()` method.  Strictly speaking it's a bug if a third-party event loop doesn't support this argument, but we don't live in an ideal world, so some will, very likely, not support `name` in 3.8. So we want to do the following:

```python
def asyncio.create_task(coro, *, name=None):
    """Schedule the execution of a coroutine object in a spawn task.

    Return a Task object.
    """
    loop = events.get_running_loop()
    task = loop.create_task(coro)

    # --- new code
    if name is not None:
        try:
            set_name = task.set_name
        except AttributeError:
            pass
        else:
            set_name(name)
    # --- end of new code

    return task
```

4. I recommend to factor out the name setting snippet into a helper function `tasks.py/_set_task_name` and use it in `loop.create_task` and `asyncio.create_task` functions.

5. We don't update the task factory API. Yes, it's suboptimal, but it's more important that this change doesn't break the world for some people.

6. You have the following snippet in the updated documentation:

   >  If *name* is ``None``, the factory must provide an automatically generated name for the task.

   This no longer applies.  asyncio Tasks will have an auto-generated name; libraries with third-party Task implementations will have to update their code to do that, eventually, but it shouldn't be a requirement, so no "must".

**Review by 1st1 — CHANGES_REQUESTED** [review]
See https://github.com/python/cpython/pull/8547#issuecomment-409245319 for more details.

**Comment by agronholm** [comment]
> loop.create_task() should be modified in such a way that it supports existing task factories.

I've already taken this into consideration. The `name` argument is passed to the task factory only if someone is trying to use the new feature and assign a name to the task. Therefore nothing will break for existing code. As for the case where the developer is trying to assign an explicit name to the task and the third party library does not support this (an unlikely scenario at the point where 3.8 has been released), a `TypeError` is raised. In my opinion, catching this and retrying without the `name` argument would also be the wrong approach. I frankly don't understand what the problem is or how adding `Task.set_name()` would fix this.

> We don't want to update the task factory API. Yes, it's suboptimal (and ugly), but it's more important that this change doesn't break the world for some people.

The same arguments apply here as in the previous quote.

> This no longer applies. asyncio Tasks will have an auto-generated name; libraries with third-party Task implementations will have to update their code to do that, eventually, but it shouldn't be a requirement, so we don't say "must".

Fair enough, I will change this to "should", okay?

**Comment by 1st1** [comment]
> The name argument is passed to the task factory only if someone is trying to use the new feature and assign a name to the task. Therefore nothing will break for existing code.

But it might break with some new code when you want to upgrade to Python 3.8. For example:

* a user has a very complex asyncio codebase;
* they use library **X** and a third-party event loop **LOOP**;
* there's a new version of **X**, which is updated and uses the new `asyncio.create_task(name='foo')` fancy API; this means that `loop.create_task()` is called with the `name` argument;
* however **LOOP** (think uvloop or tokio) hasn't yet been updated to work with 3.8, so its `loop.create_task()` **does not** support the `name` argument;
* or, alternatively, **LOOP** can use a custom `Task` implementation that doesn't have the `name` argument;
* so the user has to wait until **LOOP** (and alike dependencies) are updated to add the `name` parameter, otherwise than **cannot use X**. This can take a while, so the user will be blocked to update to Python 3.8.

The above scenario isn't strictly a case of "backwards compatibility". It's more related to how easy it is to upgrade to Python 3.8 for users who use 100s of asyncio packages (think users like Instagram or Microsoft).  My proposal makes new APIs more forgiving to third-party libraries that use low-level APIs and will not be properly upgraded for 3.8 in time.

This is a valid concern and I cannot merge the PR in its current form.

> As for the case where the developer is trying to assign an explicit name to the task and the third party library does not support this (an unlikely scenario at the point where 3.8 has been released), a TypeError is raised.

We no longer have the `Task.name` property. The current PR has only `Task.get_name()`; I'm proposing to add `Task.set_name()` (IIRC you advocated for it initially)

> I frankly don't understand what the problem is or how adding Task.set_name() would fix this.

Hopefully this comment addresses that. I understand your frustration, but contributing seemingly simple new APIs like this one had never been a simple task; asyncio is too big now and there are too many lines of code and production environments that depend on it simply working in new Python releases.

**Comment by 1st1** [comment]
The key issue here is that any forward compatibility pain in this particular case can be easily avoided. @asvetlov what's your opinion on this?

**Comment by 1st1** [comment]
> So what you're saying is that if someone is using a library that explicitly assigns names to task but via an event loop implementation that does not yet support task names, it will fail. But the same will happen if said library uses Task.set_name() instead of passing it to create_task(). The only difference I see here is which exception they will get: TypeError vs AttributeError.

That's fair, and same applies to *any new* asyncio APIs, you are right.
 
In this particular case I'm really concerned with breaking custom task factories, as this change forces you to **update your existing code** even if your own code does not use task names, but some dependency does. This is different from adding a new method to asyncio, where you **don't update your code** when some dependancy starts using new features.

IOW I'm trying to argue here that we can *completely* avoid any potential future compatibility problem in this very case with some, admittedly ugly, workaround in asyncio.

**Comment by agronholm** [comment]
> IOW I'm trying to argue here that we can completely avoid any potential future compatibility problem in this very case with some, admittedly ugly, workaround in asyncio.

Sorry to drag this on but I don't understand how this workaround would solve this problem. Suppose that we refrain from modifying the task factory API but add `Task.set_name()`, a dependency which tries to set the task name would still get an `AttributeError` since the custom task factory could still be producing `Task` instances which don't have the `set_name()` method. So what am I missing?

**Comment by 1st1** [comment]
> Suppose that we refrain from modifying the task factory API but add Task.set_name(), a dependency which tries to set the task name would still get an AttributeError since the custom task factory could still be producing Task instances which don't have the set_name() method. So what am I missing?

I think that it's far less likely that everybody starts using `Task.set_name()`. I think it's *more likely* that *some* dependancy might start using `asyncio.create_task(coro, name='something')` and your application breaks in a non-trivial way because you use Tokio event loop that hasn't yet implemented `Task.set_name()`. Or because something has an old task factory that will take a while to get fixed.

The more I think about this, the less I want to change the signature of task factories. There's no need for that.

**Comment by agronholm** [comment]
While I am still completely unconvinced by your assessment and reasoning, I will go ahead and implement `Task.set_name()` so we can proceed with merging this PR.

**Comment by 1st1** [comment]
> While I am still completely unconvinced by your assessment and reasoning, I will go ahead and implement Task.set_name() so we can proceed with merging this PR.

Thank you.

**Comment by asvetlov** [comment]
I agree with @1st1 argumentation

File: Doc/library/asyncio-task.rst
Comment (1st1) [inline Doc/library/asyncio-task.rst]: I'm confused... where's the `name` argument? Please take another look at https://github.com/python/cpython/pull/8547#issuecomment-409245319.

Perhaps the problem is that we mean different things by "task factories":

* I mean custom user factories set by `loop.set_task_factory()` (which `loop.create_task()` later calls)
* You seem to mean `loop.create_task()`.

Maybe this is where the confusion is coming from.  Please re-read https://github.com/python/cpython/pull/8547#issuecomment-409245319.

Comment (agronholm) [reply Doc/library/asyncio-task.rst]: You seem to be suggesting that we ignore any `AttributeError` produced by a misbehaving task factory. This is something I've also considered and then rejected. Sweeping problems like that under the carpet is something I absolutely want to avoid. The best solution would, in my opinion, be to raise a user friendly error message explaining why setting a custom name for the task failed.

Comment (1st1) [reply Doc/library/asyncio-task.rst]: > Sweeping problems like that under the carpet is something I absolutely want to avoid. The best solution would, in my opinion, be to raise a user friendly error message explaining why setting a custom name for the task failed.

Well, since task names are not critical in *any way* for asyncio programs I tend to disagree here. They are just here to simplify debugging.

I'm OK with issuing a warning though, like "some.lib.name.Task does not have a set_name() method".  We can later (in Python 3.9 or 4?) convert this into an error.

Comment (agronholm) [reply Doc/library/asyncio-task.rst]: A warning seems to be a reasonable compromise here. I will go add the `name` argument back.

Comment (1st1) [reply Doc/library/asyncio-task.rst]: Great, thanks!

Comment (agronholm) [reply Doc/library/asyncio-task.rst]: For some reason I had missed parts of your commentary. It now makes a lot more sense to me. With the addition of the warning, I think we're finally on the same page. The one change I still want to make is to keep the `name` keyword argument in the stdlib Task constructor to avoid needlessly generating a task name if one was explicitly provided. Agreed?

Comment (1st1) [reply Doc/library/asyncio-task.rst]: > For some reason I had missed parts of your commentary. It now makes a lot more sense to me. With the addition of the warning, I think we're finally on the same page.

Great, I had a feeling that something is off, but couldn't pinpoint that we mean different things when we discuss "task factories" :)

> The one change I still want to make is to keep the name keyword argument in the stdlib Task constructor to avoid needlessly generating a task name if one was explicitly provided.

OK by me.

**Comment by agronholm** [comment]
I have made the requested changes; please review again.

**Comment by agronholm** [comment]
@1st1 is this okay now? It is missing the warning I talked about, not 100% sure we need it. Others' opinions are welcome.

File: Doc/library/asyncio-eventloop.rst
Comment (1st1) [inline Doc/library/asyncio-eventloop.rst]: I'd change to "If *name* argument is provided, ..."

File: Modules/_asynciomodule.c
Comment (1st1) [inline Modules/_asynciomodule.c]: Please change 

```c
} else {
``` 

to 

```c
}
else if (!PyUnicode_Check(name)) {
```

File: Lib/asyncio/base_events.py
Comment (1st1) [inline Lib/asyncio/base_events.py:396]: This line should be unindented or you should pass the name to `task.Task(..)` few lines above.  Please add a test that Tasks get custom names when created via `loop.create_task(name=..)` with a standard factory.

**Comment by agronholm** [comment]
I have made the requested changes; please review again.

File: Lib/asyncio/tasks.py
Comment (1st1) [inline Lib/asyncio/tasks.py]: nit: I've recently updated asyncio to use f-strings. Please change this one to either an f-string or to a `.format()` call (just for maintaining consistency of the codebase).

File: Lib/test/test_asyncio/test_base_events.py
Comment (1st1) [inline Lib/test/test_asyncio/test_base_events.py]: Please also add `test_create_named_task_with_default_factory`. That would test a bug we fixed in this PR when the default task factory didn't pass the `name=name` to `Task()`.

**Comment by 1st1** [comment]
Almost there. Please fix the two remaining nit comments and I'll merge this.

**Review by 1st1 — APPROVED** [review]

File: Modules/_asynciomodule.c
Comment (1st1) [inline Modules/_asynciomodule.c]: Oups, need to fix this to `return NULL`!

**Review by 1st1 — CHANGES_REQUESTED** [review]
Just spotted an error, a `PyObject*` function returns `-1`.  Please also check that the C compiler generates 0 warnings for your patch in `_asynciomodule.c`.

File: Modules/_asynciomodule.c
Comment (agronholm) [reply Modules/_asynciomodule.c]: With `make clean;  CFLAGS=-Wno-cast-function-type make`, compilation of the `_asyncio` extension produced no warnings after the patch.

**Comment by agronholm** [comment]
I have made the requested changes; please review again.

**Review by 1st1 — APPROVED** [review]

**Comment by 1st1** [comment]
Merged. Thanks for much for pushing this through, Alex!

**Pull request merged**
